lib/knapsack_pro/config/temp_files.rb

```diff
@@ -0,0 +1,46 @@
+module KnapsackPro
+  module Config
+    class TempFiles
+      def self.temp_directory_path
```

Comment:
Have you considered separating the ensuring the temp directory exists from returning the path? Or maybe memoizing the result or sth?

Do you think there is a way to do this check once at some point, or would it always be succeptible to errors?

The reason I'm wondering about this is that it seems that we are using this `.temp_directory_path` method in many places to just get the correct path, but each time we do this it would now trigger an I/O read operation. Would be great to avoid that if it's avoidable.

Reply:
I was thinking about it and I decided that checking if file exists on the disk is good enough. We already do some IO operations in queue mode like saving reports and it shouldn't have significant impact on performance.

If I add memorization to class method then this causes random tests failures depending on the order of tests executed because class method memorization is preserved for each call of class method.

For instance I could do:

```ruby
      def self.temp_directory_path
        @temp_directory_path ||=
          begin
            temp_files = new
            temp_files.ensure_temp_directory_exists
            temp_files.temp_directory_path
          end
      end
```

and then when I run tests I get randomly different tests failures:

```
[14:47:44] artur@Artur-MBP: ~/Documents/github/knapsack-pro/knapsack_pro-ruby temporary-files-directory! ruby-3.0.2
$ rspec ./spec/knapsack_pro/config/temp_files_spec.rb

Randomized with seed 42508

KnapsackPro::Config::TempFiles
  .temp_directory_path
    .gitignore file has content
    returns temporary directory path
    creates .gitignore file (FAILED - 1)

Failures:

  1) KnapsackPro::Config::TempFiles.temp_directory_path creates .gitignore file
     Failure/Error: expect(File.exist?(gitignore_file_path)).to be true

       expected true
            got false
     # ./spec/knapsack_pro/config/temp_files_spec.rb:14:in `block (3 levels) in <top (required)>'
     # /Users/artur/.rvm/gems/ruby-3.0.2/gems/webmock-3.13.0/lib/webmock/rspec.rb:37:in `block (2 levels) in <top (required)>'

Finished in 0.02793 seconds (files took 0.53206 seconds to load)
3 examples, 1 failure

Failed examples:

rspec ./spec/knapsack_pro/config/temp_files_spec.rb:11 # KnapsackPro::Config::TempFiles.temp_directory_path creates .gitignore file

Randomized with seed 42508


[14:47:46] artur@Artur-MBP: ~/Documents/github/knapsack-pro/knapsack_pro-ruby temporary-files-directory! ruby-3.0.2
$ rspec ./spec/knapsack_pro/config/temp_files_spec.rb

Randomized with seed 56734

KnapsackPro::Config::TempFiles
  .temp_directory_path
    returns temporary directory path
    .gitignore file has content (FAILED - 1)
    creates .gitignore file (FAILED - 2)

Failures:

  1) KnapsackPro::Config::TempFiles.temp_directory_path .gitignore file has content
     Failure/Error: expect(File.read(gitignore_file_path)).to include '# This directory is used by knapsack_pro gem for storing temporary files during tests runtime'

     Errno::ENOENT:
       No such file or directory @ rb_sysopen - .knapsack_pro/.gitignore
     # ./spec/knapsack_pro/config/temp_files_spec.rb:19:in `read'
     # ./spec/knapsack_pro/config/temp_files_spec.rb:19:in `block (3 levels) in <top (required)>'
     # /Users/artur/.rvm/gems/ruby-3.0.2/gems/webmock-3.13.0/lib/webmock/rspec.rb:37:in `block (2 levels) in <top (required)>'

  2) KnapsackPro::Config::TempFiles.temp_directory_path creates .gitignore file
     Failure/Error: expect(File.exist?(gitignore_file_path)).to be true

       expected true
            got false
     # ./spec/knapsack_pro/config/temp_files_spec.rb:14:in `block (3 levels) in <top (required)>'
     # /Users/artur/.rvm/gems/ruby-3.0.2/gems/webmock-3.13.0/lib/webmock/rspec.rb:37:in `block (2 levels) in <top (required)>'

Finished in 0.01396 seconds (files took 0.39825 seconds to load)
3 examples, 2 failures

Failed examples:

rspec ./spec/knapsack_pro/config/temp_files_spec.rb:17 # KnapsackPro::Config::TempFiles.temp_directory_path .gitignore file has content
rspec ./spec/knapsack_pro/config/temp_files_spec.rb:11 # KnapsackPro::Config::TempFiles.temp_directory_path creates .gitignore file

Randomized with seed 56734
```

I could disable memorization with custom argument `def self.temp_directory_path(cached: true)`. For the test environment I could use`cache: false` but I'm not sure if this is good idea to change method behaviour only to make tests pass :/